presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java

```diff
@@ -0,0 +1,227 @@
+
+    public PushPartialAggregationThroughExchange(FunctionRegistry functionRegistry)
+    {
+        this.functionRegistry = functionRegistry;
+    }
+
+    @Override
+    public Pattern getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Context context)
+    {
+        AggregationNode aggregationNode = (AggregationNode) node;
+
+        PlanNode childNode = context.getLookup().resolve(aggregationNode.getSource());
+        if (!(childNode instanceof ExchangeNode)) {
+            return Optional.empty();
+        }
+
+        ExchangeNode exchangeNode = (ExchangeNode) childNode;
+
+        boolean decomposable = aggregationNode.isDecomposable(functionRegistry);
+
+        if (aggregationNode.getStep().equals(SINGLE) &&
+                aggregationNode.hasEmptyGroupingSet() &&
+                aggregationNode.hasNonEmptyGroupingSet()) {
+            checkState(
+                    decomposable,
+                    "Distributed aggregation with empty grouping set requires partial but functions are not decomposable");
+            return Optional.of(split(aggregationNode, context));
+        }
+
+        if (!decomposable) {
+            return Optional.empty();
+        }
+
+        // partial aggregation can only be pushed through exchange that doesn't change
+        // the cardinality of the stream (i.e., gather or repartition)
+        if ((exchangeNode.getType() != GATHER && exchangeNode.getType() != REPARTITION) ||
+                exchangeNode.getPartitioningScheme().isReplicateNullsAndAny()) {
+            return Optional.empty();
+        }
+
+        if (exchangeNode.getType() == REPARTITION) {
+            // if partitioning columns are not a subset of grouping keys,
+            // we can't push this through
+            List<Symbol> partitioningColumns = exchangeNode.getPartitioningScheme()
+                    .getPartitioning()
+                    .getArguments()
+                    .stream()
+                    .filter(Partitioning.ArgumentBinding::isVariable)
+                    .map(Partitioning.ArgumentBinding::getColumn)
+                    .collect(Collectors.toList());
+
+            if (!aggregationNode.getGroupingKeys().containsAll(partitioningColumns)) {
+                return Optional.empty();
+            }
+        }
+
+        // currently, we only support plans that don't use pre-computed hash functions
+        if (aggregationNode.getHashSymbol().isPresent() || exchangeNode.getPartitioningScheme().getHashColumn().isPresent()) {
+            return Optional.empty();
+        }
+
+        switch (aggregationNode.getStep()) {
+            case SINGLE:
+                // Split it into a FINAL on top of a PARTIAL and
+                // reprocess the resulting plan to push the partial
+                // below the exchange (see case below).
+                return Optional.of(split(aggregationNode, context));
```

Comment:
Without splitting tests fail. I prefer to leave `PushPartialAggregationThroughExchange` as is for now since it is even more complex than `PushPartialAggregationThroughJoin`